File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/9c5901eae3b739bcb095f3a3a6d4b8b0.py
import sys
import random
from math import acos

INF = int(1e9)
INF64 = int(1e18)
MOD = 998244353
EPS = 1e-9
PI = acos(-1.0)

N = 300 * 1000 + 13

n = 0
g = [[] for _ in range(N)]

def read():
    global n
    try:
        n = int(sys.stdin.readline().strip())
    except ValueError:
        return False
    for i in range(n):
        g[i].clear()
    for _ in range(n - 1):
        v, u = map(int, sys.stdin.readline().strip().split())
        v -= 1
        u -= 1
        g[v].append(u)
        g[u].append(v)
    return True

def add(a, b):
    a += b
    if a >= MOD:
        a -= MOD
    if a < 0:
        a += MOD
    return a

def mul(a, b):
    return a * b % MOD

dp = [[0] * 3 for _ in range(N)]

def dfs(v, p=-1):
    dp[v][0] = 1
    dp[v][1] = 1
    vals = []
    init = []
    for u in g[v]:
        if u != p:
            dfs(u, v)
            vals.append(add(dp[u][1], add(dp[u][2], dp[u][2])))
            init.append(add(dp[u][0], dp[u][1]))
            dp[v][0] = mul(dp[v][0], add(dp[u][1], add(dp[u][2], dp[u][2])))
            dp[v][1] = mul(dp[v][1], add(dp[u][1], dp[u][2]))
    
    dp[v][0] = add(dp[v][0], -dp[v][1])
    
    pr = [1]
    su = [1]
    for it in vals:
        pr.append(mul(pr[-1], it))
    vals.reverse()
    for it in vals:
        su.append(mul(su[-1], it))
    vals.reverse()
    su.reverse()
    
    dp[v][2] = 0
    for i in range(len(vals)):
        dp[v][2] = add(dp[v][2], mul(init[i], mul(pr[i], su[i + 1])))

def solve():
    dfs(0)
    print(add(dp[0][1], dp[0][2]))

def main():
    input_file = open("input.txt", "r")
    sys.stdin = input_file

    while read():
        solve()

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Java/9c5901eae3b739bcb095f3a3a6d4b8b0.java
import java.util.*;
import java.io.*;

public class 9c5901eae3b739bcb095f3a3a6d4b8b0xcodeeval_processed_4000.json {
    static final int INF = (int) 1e9;
    static final long INF64 = (long) 1e18;
    static final int MOD = 998244353;
    static final double EPS = 1e-9;
    static final double PI = Math.acos(-1.0);
    static final int N = 300 * 1000 + 13;

    static int n;
    static List<Integer>[] g = new ArrayList[N];
    static int[][] dp = new int[N][3];

    static {
        for (int i = 0; i < N; i++) {
            g[i] = new ArrayList<>();
        }
    }

    static boolean read() {
        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            n = Integer.parseInt(br.readLine());
            for (int i = 0; i < n; i++) {
                g[i].clear();
            }
            for (int i = 0; i < n - 1; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                int v = Integer.parseInt(st.nextToken()) - 1;
                int u = Integer.parseInt(st.nextToken()) - 1;
                g[v].add(u);
                g[u].add(v);
            }
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    static int add(int a, int b) {
        a += b;
        if (a >= MOD) a -= MOD;
        if (a < 0) a += MOD;
        return a;
    }

    static int mul(int a, int b) {
        return (int) ((1L * a * b) % MOD);
    }

    static void dfs(int v, int p) {
        dp[v][0] = 1;
        dp[v][1] = 1;
        List<Integer> vals = new ArrayList<>();
        List<Integer> init = new ArrayList<>();
        for (int u : g[v]) {
            if (u != p) {
                dfs(u, v);
                vals.add(add(dp[u][1], add(dp[u][2], dp[u][2])));
                init.add(add(dp[u][0], dp[u][1]));
                dp[v][0] = mul(dp[v][0], add(dp[u][1], add(dp[u][2], dp[u][2])));
                dp[v][1] = mul(dp[v][1], add(dp[u][1], dp[u][2]));
            }
        }
        dp[v][0] = add(dp[v][0], -dp[v][1]);

        List<Integer> pr = new ArrayList<>(Collections.singletonList(1));
        List<Integer> su = new ArrayList<>(Collections.singletonList(1));
        for (int it : vals) {
            pr.add(mul(pr.get(pr.size() - 1), it));
        }
        Collections.reverse(vals);
        for (int it : vals) {
            su.add(mul(su.get(su.size() - 1), it));
        }
        Collections.reverse(vals);
        Collections.reverse(su);

        dp[v][2] = 0;
        for (int i = 0; i < vals.size(); i++) {
            dp[v][2] = add(dp[v][2], mul(init.get(i), mul(pr.get(i), su.get(i + 1))));
        }
    }

    static void solve() {
        dfs(0, -1);
        System.out.println(add(dp[0][1], dp[0][2]));
    }

    public static void main(String[] args) {
        try {
            BufferedReader br = new BufferedReader(new FileReader("input.txt"));
            PrintStream out = new PrintStream(new FileOutputStream("output.txt"));
            System.setOut(out);
            long tt = System.currentTimeMillis();

            while (read()) {
                solve();
                System.err.println("TIME = " + (System.currentTimeMillis() - tt));
                tt = System.currentTimeMillis();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
